### PR TITLE
feat: bun related docs (@stonega)

### PR DESCRIPTION
**Pull request by MasedMSD**
Originally [PR](https://github.com/grammyjs/website/pull/1105) was created by @stonega, but it looks like he's not going to continue with this PR. So, I think we should mention Bun as a possible way to work with a bot, even though Bun has a large number of issues. Also Bun finally supports [conversations plugin](https://grammy.dev/plugins/conversations) and a decent part of Russian language chat users use it for their projects.

I made some changes and added Bun mentions to the hosting files documentation

### CODE REVIEW

File: site/docs/guide/getting-started.md [hunk 1]
```diff
@@ -150,6 +150,90 @@ in your terminal before you execute `node bot.js`.
...
+// This will connect to the Telegram servers and wait for messages.
+
+// Start the bot.
+bot.start();
+```
+
+You can now run the bot by executing
+
+```sh
+bun run bot.ts
```
Comment (MasedMSD) [inline site/docs/guide/getting-started.md:218]: Didn't changed that
```
bun run index.ts
```

File: site/docs/guide/introduction.md [hunk 2]
```diff
@@ -173,6 +173,34 @@ You can stop the bot again with `Ctrl+C`.
...
+```sh
+mkdir ./my-bot
+cd ./my-bot
+code .
+```
+
+Then:
+
+1. Run `bun init --yes` in your terminal to initialize the project.
+2. Create a source file `bot.ts` with TypeScript code inside the project.
```
Comment (MasedMSD) [inline site/docs/guide/introduction.md:198]: Should change `bot.ts` to `index.ts` which bun created by default when using `bun init --yes`

File: site/docs/guide/getting-started.md [hunk 3]
```diff
@@ -150,6 +150,90 @@ in your terminal before you execute `node bot.js`.
...
+Create a new project and install the `grammy` package.
+Do this by opening a terminal and typing:
+
+```sh
+# Create a new directory and change into it.
+mkdir my-bot
+cd my-bot
+
+# Run bun init to scaffold a new project.
+bun init --yes
```
Comment (KnorpelSenf) [inline site/docs/guide/getting-started.md:167]: Does Bun really require a setup script?

File: site/docs/guide/getting-started.md (hunk 1 above)
Comment (KnorpelSenf) [reply site/docs/guide/getting-started.md:218]: Why not?

Comment (KnorpelSenf) [reply site/docs/guide/getting-started.md:218]: Actually, it would be much better to create a `bot.ts` file instead of `index.ts`. It makes no sense to do this inconsistently for Bun.

File: site/docs/guide/introduction.md (hunk 2 above)
Comment (KnorpelSenf) [reply site/docs/guide/introduction.md:198]: How can this be fixed?

File: site/docs/guide/getting-started.md (hunk 1 above)
Comment (MasedMSD) [reply site/docs/guide/getting-started.md:218]: > Why not?

Just forgot about it. These are comments addressed to me in the morning 😶‍🌫️

Comment (MasedMSD) [reply site/docs/guide/getting-started.md:218]: > Actually, it would be much better to create a `bot.ts` file instead of `index.ts`. It makes no sense to do this inconsistently for Bun.

Hm, you're right. If all the examples use `bot.ts`, it would be convenient to leave this for association

File: site/docs/guide/introduction.md (hunk 2 above)
Comment (MasedMSD) [reply site/docs/guide/introduction.md:198]: > How can this be fixed?

Just change filename

File: site/docs/guide/getting-started.md (hunk 3 above)
Comment (MasedMSD) [reply site/docs/guide/getting-started.md:167]: Not really, Bun just uses it for creating default files like README, package.json, tsconfig and index.ts.

I don't know about older versions, but the new ones add the package without init

**Comment by MasedMSD** [comment]
Should we leave `index.ts` or should we tell the user to create a `bot.ts` file each time in order for the NodeJS and Deno examples to be the same?

File: site/docs/guide/getting-started.md (hunk 1 above)
Comment (KnorpelSenf) [reply site/docs/guide/getting-started.md:218]: Do we have to use the init script then? Maybe we can just create the files manually or use `npm init`?

File: site/docs/guide/introduction.md (hunk 2 above)
Comment (KnorpelSenf) [reply site/docs/guide/introduction.md:198]: Wait so there's no way to skip creating the file, or so create it with a different name?

File: site/docs/guide/getting-started.md (hunk 1 above)
Comment (MasedMSD) [reply site/docs/guide/getting-started.md:218]: Yeah, why not. If the Bun has its own initialization that looks more convenient than npm, why we shouldn't use it

File: site/docs/guide/introduction.md (hunk 2 above)
Comment (MasedMSD) [reply site/docs/guide/introduction.md:198]: I don't guess. Bun by default create index.ts with content:

```ts
console.log("Hello via Bun!");
```

File: site/docs/guide/getting-started.md (hunk 3 above)
Comment (KnorpelSenf) [reply site/docs/guide/getting-started.md:167]: tsconfig can't be created by bun, unless it installs and runs `tsc`? Because the rest of the website depends on having the TypeScript default config and I don't think I want to trust bun on making sure that these two files are always in sync, so we have to tell people to run `tsc` themselves.

Comment (MasedMSD) [reply site/docs/guide/getting-started.md:167]: > tsconfig can't be created by bun, unless it installs and runs `tsc`? Because the rest of the website depends on having the TypeScript default config and I don't think I want to trust bun on making sure that these two files are always in sync, so we have to tell people to run `tsc` themselves.

I don't quite understand the question. Bun does not use tsc for compilation. It runs Typescript immediately

File: site/docs/guide/getting-started.md (hunk 1 above)
Comment (KnorpelSenf) [reply site/docs/guide/getting-started.md:218]: Because it clearly isn't more convenient, it creates the wrong files

Comment (MasedMSD) [reply site/docs/guide/getting-started.md:218]: What you mean by `creates the wrong files`

File: site/docs/guide/introduction.md (hunk 2 above)
Comment (KnorpelSenf) [reply site/docs/guide/introduction.md:198]: I don't see how we need that. Note that we also don't run `deno init` for the same reason.

Comment (MasedMSD) [reply site/docs/guide/introduction.md:198]: > I don't see how we need that. Note that we also don't run `deno init` for the same reason.

This is to show that we are starting a clean project. Of course, we may not init the project and start a ready-made one.

File: site/docs/guide/getting-started.md (hunk 3 above)
Comment (KnorpelSenf) [reply site/docs/guide/getting-started.md:167]: This is incorrect. It should be this:
```json
{
  "compilerOptions": {
    "target": "es2016",
    "module": "commonjs",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true
  }
}
```

Comment (MasedMSD) [reply site/docs/guide/getting-started.md:167]: Why?

File: site/docs/guide/getting-started.md (hunk 1 above)
Comment (KnorpelSenf) [reply site/docs/guide/getting-started.md:218]: The files it creates have the wrong content.
- `index.ts` should actually be named `bot.ts`
- `bot.ts` should actually different content
- `tsconfig.json` is wrong
- we have not even talked about package.json, will this also be a problem?

File: site/docs/guide/introduction.md (hunk 2 above)
Comment (KnorpelSenf) [reply site/docs/guide/introduction.md:198]: Since almost everything `bun init` does is wrong for our use case, I don't understand why it would be useful. The docs would be:
1. Run `bun init`
2. Delete all files it created
3. Create correct files

File: site/docs/guide/getting-started.md (hunk 1 above)
Comment (MasedMSD) [reply site/docs/guide/getting-started.md:218]: - `index.ts` should actually be named `bot.ts` - not a problem.
- `bot.ts` should actually have different content - not a problem again. Of course user should write default code for bot launching
- `tsconfig.json` is wrong - what's wrong with it?
- we have not even talked about package.json, will this also be a problem? - Not actually. package.json is:

```JSON
{
  "name": "test",
  "module": "index.ts",
  "type": "module",
  "dependencies": {
    "grammy": "^1.30.0"
  },
  "devDependencies": {
    "@types/bun": "latest"
  },
  "peerDependencies": {
    "typescript": "^5.0.0"
  }
}
```

Comment (MasedMSD) [reply site/docs/guide/getting-started.md:218]: image

File: site/docs/guide/introduction.md (hunk 2 above)
Comment (MasedMSD) [reply site/docs/guide/introduction.md:198]: It's not deleting already created files

File: site/docs/guide/getting-started.md (hunk 3 above)
Comment (KnorpelSenf) [reply site/docs/guide/getting-started.md:167]: Because the rest of the docs require it. Every single of the ~700 code examples are tested on both Deno and Node and they work with the config created in the getting started guide. So we either make sure that Bun is compatible with the Node examples, or else we will have to re-create all code examples with a Bun tab. Even if it turns out to the mostly the same in most cases, we have to least test everything.

I don't care about bun and I don't want to spend time on this. I am not going install it. I also want to be able to edit the docs myself, and write new docs. This is not possible if I have to test everything on bun first. I won't accept any contributions that prevent me from changing the docs in the future, or that force me to ask for help for every single edit.

The bottom line is that the bun setup MUST work with the existing Node code. It must type-check with the same TS config, and it must run correctly.

Comment (MasedMSD) [reply site/docs/guide/getting-started.md:167]: > The bottom line is that the bun setup MUST work with the existing Node code. It must type-check with the same TS config, and it must run correctly.

I understand your indignation about this runtime, but I have not had any problems with it before when working with it.

Comment (KnorpelSenf) [reply site/docs/guide/getting-started.md:167]: Actually, nobody will test if it runs correctly, people are just going to watch it fail when they use Bun. But at least we can make sure that it type-checks.

File: site/docs/guide/getting-started.md (hunk 1 above)
Comment (KnorpelSenf) [reply site/docs/guide/getting-started.md:218]: Why is TS a peer dep? It should be a dev dep.

File: site/docs/guide/introduction.md (hunk 2 above)
Comment (KnorpelSenf) [reply site/docs/guide/introduction.md:198]: Read again

File: site/docs/guide/getting-started.md (hunk 3 above)
Comment (MasedMSD) [reply site/docs/guide/getting-started.md:167]: > This is incorrect. It should be this:
> 
> ```json
> {
>   "compilerOptions": {
>     "target": "es2016",
>     "module": "commonjs",
>     "esModuleInterop": true,
>     "forceConsistentCasingInFileNames": true,
>     "strict": true,
>     "skipLibCheck": true
>   }
> }
> ```

I'll check all examples with Bun if you want so
image

Comment (KnorpelSenf) [reply site/docs/guide/getting-started.md:167]: > > The bottom line is that the bun setup MUST work with the existing Node code. It must type-check with the same TS config, and it must run correctly.
> 
> I understand your indignation about this runtime, but I have not had any problems with it before when working with it.

Sure but you also did not test everything

**Comment by KnorpelSenf** [comment]
I have now wasted more time on this than I ever wanted, maybe I'll get back to this in a week or so

File: site/docs/guide/getting-started.md (hunk 1 above)
Comment (MasedMSD) [reply site/docs/guide/getting-started.md:218]: Bun adds a tiзуscript to Peer Deps, but honestly I haven't tried to figure out the behaviour of that. We could move it to dev deps and nothing would change.

**Review by KnorpelSenf — CHANGES_REQUESTED** [review]
The bottom line of this is still that whatever we do with the tutorial, the output must look exactly like https://github.com/grammyjs/website/pull/1129/files#diff-a3f07fb94b91719ead29e589d98c27db44c1f14974480c5960176a2c2c35b6efR56-R66 and the file contents need to be the same, too. I still do not really care how that is achieved, the point is that as long as the bun guide does not lead to the correct setup, then this cannot be merged.

**Comment by MasedMSD** [comment]
I'll get into this page in more detail tomorrow

**Comment by KnorpelSenf** [comment]
Remember to merge `main` in order to fix CI

**Comment by KnorpelSenf** [comment]
Scratch that, it already happened

**Comment by KnorpelSenf** [comment]
I even did it myself \*facepalm\*

**Review by KnorpelSenf — COMMENTED** [review]
This still does not create the same TS config as `tsc --init` if I understand this correctly

**Comment by MasedMSD** [comment]
Unfortunately we can't fix it. We will have to ask the user to change it manually every time.
But I don't understand the point of this, because everything works with the current tsconfig O.O

**Comment by KnorpelSenf** [comment]
Yeah but I don't want to go through all the code examples and make sure that they compile even with the non-standard TS config that bun introduces. It can be fixed very easily by running `tsc --init` in the getting started guide.

File: site/docs/guide/getting-started.md [hunk 4]
```diff
@@ -150,6 +150,116 @@ in your terminal before you execute `node bot.js`.
...
+```sh
+# Create a new directory and change into it.
+mkdir my-bot
+cd my-bot
+```
+
+Next step you should initialise your project.
+
+```sh
+# Run bun init to scaffold a new project.
```
Comment (winstxnhdw) [inline site/docs/guide/getting-started.md:170]:
```suggestion
# Run bun init to scaffold a new project.
tsc —init
```

**Review by winstxnhdw — CHANGES_REQUESTED** [review]

File: site/docs/guide/getting-started.md (hunk 4 above)
Comment (KnorpelSenf) [reply site/docs/guide/getting-started.md:170]: Nice. Do we also need a `bun add -D typescript` somewhere in there, so that people have `tsc` ready? We might also want to use `bunx`.

Comment (MasedMSD) [reply site/docs/guide/getting-started.md:170]: We can't init typescript before `bun init`.
`bun init` creates `package.json` and there could possibly be problems with that.

Comment (MasedMSD) [reply site/docs/guide/getting-started.md:170]: We can just use bunx for that btw.
Like

```markdown
# Use bun to init Typescript config
bunx tsc --init
# Run bun init to scaffold a new project.
bun init
```

Comment (KnorpelSenf) [reply site/docs/guide/getting-started.md:170]: Right, so adding TypeScript would be done after these two commands? That makes sense

**Review by quadratz — CHANGES_REQUESTED** [review]

File: site/docs/guide/getting-started.md (hunk 4 above)
Comment (quadratz) [reply site/docs/guide/getting-started.md:170]: Wait, `bun init`  already includes a `tsconfig.json`, as explained in https://bun.sh/docs/typescript. I don't think this step is necessary, no?

Comment (winstxnhdw) [reply site/docs/guide/getting-started.md:170]: > Wait, `bun init` already includes a `tsconfig.json`, as explained in https://bun.sh/docs/typescript. I don't think this step is necessary, no?

See the earlier comments about keeping tsconfig consistent.

**Comment by quadratz** [comment]
> @KnorpelSenf : Yeah but I don't want to go through all the code examples and make sure that they compile even with the non-standard TS config that bun introduces. It can be fixed very easily by running tsc --init in the getting started guide.
> > @MasedMSD: But I don't understand the point of this, because everything works with the current tsconfig O.O

I am agree with @MasedMSD. I don't see any problem with the [default generated tsconfig.json](https://bun.sh/docs/typescript#suggested-compileroptions). Pretty sure it works well with our examples.

File: site/docs/guide/introduction.md [hunk 5]
```diff
@@ -66,7 +66,7 @@ You will get to know them as you go.
 
 ## Prerequisites to Getting Started
 
-> Skip the rest of this page if you already know how to develop a Deno or a Node.js application, and [get started](./getting-started).
+> Skip the rest of this page if you already know how to develop Deno, Bun or Node.js application, and [get started](./getting-started).
```
Comment (quadratz) [inline site/docs/guide/introduction.md:69]:
```suggestion
> Skip the rest of this page if you already know how to develop Deno, Bun, or Node.js application, and [get started](./getting-started).
```

File: site/docs/guide/introduction.md [hunk 6]
```diff
@@ -173,6 +173,47 @@ You can stop the bot again with `Ctrl+C`.
 Ready?
 [Get started](./getting-started#getting-started-on-deno)! :robot:
 
+### Prerequisites for Bun
+
+Bun is a new JavaScript runtime, which is another good choice to build bot, like Deno, you are going to write your bot in TypeScript.
```
Comment (quadratz) [inline site/docs/guide/introduction.md:178]:
```suggestion
Bun is a new JavaScript runtime, which is another good choice to build bot.
Like Deno, you are going to write your bot in TypeScript.
```

**Review by quadratz — CHANGES_REQUESTED** [review]

**Comment by KnorpelSenf** [comment]
> > @KnorpelSenf : Yeah but I don't want to go through all the code examples and make sure that they compile even with the non-standard TS config that bun introduces. It can be fixed very easily by running tsc --init in the getting started guide.
> > > @MasedMSD: But I don't understand the point of this, because everything works with the current tsconfig O.O
> 
> I am agree with @MasedMSD. I don't see any problem with the [default generated tsconfig.json](https://bun.sh/docs/typescript#suggested-compileroptions). Pretty sure it works well with our examples.

Then please go through them and verify this. Also, please test all future changes to the docs for bun compatibility. I don't care what the config file contains, I just won't double the time I spend on code examples just to make a few bun users happy.

**Comment by MasedMSD** [comment]
Fine, so we going to use my [solution](https://github.com/grammyjs/website/pull/1129#discussion_r1829563911)?

**Comment by KnorpelSenf** [comment]
Only as soon as we find somebody who actually puts in the work and makes sure that the code examples work. I don't want to merge docs that don't type-check because then I'm gonna be the one who has to deal with all the complaints.

If somebody is willing to
- make sure that the current docs type-check, and
- make sure that all future docs type-check 

then we can ship docs that use a new TS config setup.

**Comment by MasedMSD** [comment]
@KnorpelSenf It would be useful to create some package for easy unit testing of the bot.
I don't think that's possible given the flood handling by telegram

**Comment by KnorpelSenf** [comment]
Is that in any way related to the current pull request?

**Comment by MasedMSD** [comment]
I mean that we should have some kind of tests to make sure that the bot and the code work. For example, as for the situation now, when we can't be sure that bun will pass all type-checks

**Comment by winstxnhdw** [comment]
> I mean that we should have some kind of tests to make sure that the bot and the code work. For example, as for the situation now, when we can't be sure that bun will pass all type-checks

It’ll be hard to write integration tests for Bun and grammY because the existing test suite is in Deno. We’d have to spend time instrumenting some Docker CI test just specifically for Bun.

**Comment by KnorpelSenf** [comment]
There's no need to actually run any tests, why do you think this? I can't follow

**Comment by MasedMSD** [comment]
@KnorpelSenf still confused. We don't gonna use [this](https://github.com/grammyjs/website/pull/1129#discussion_r1829563911)?

**Comment by KnorpelSenf** [comment]
I don't have strong opinions on what exactly you're gonna do. I just need you to make sure that the code examples are correct. It doesn't matter to me if you achieve this by making people use the setup that is known to work, or if you want to test all present and future code snippets against a new setup.

**Comment by MasedMSD** [comment]
What I mean is that this command generates the typescript config that we use in all the examples. I think we should add Bun, but if there are any problems (which I'm not sure about at all, Bun is quite stable) we should deal with them as they come up.

**Comment by KnorpelSenf** [comment]
So is there still a question? If yes then I don't understand it

**Comment by MasedMSD** [comment]
I mean, this PR has been hanging around for a long time and we need to do something about it. Either put it on hold or get on with it

**Comment by KnorpelSenf** [comment]
I don't care if you use the command or not and I'm not opposed to mentioning bun support

**Comment by MasedMSD** [comment]
> I'm not opposed to mentioning bun support

Great, I'll take care of that PR

**Comment by KnorpelSenf** [comment]
> > I'm not opposed to mentioning bun support
> 
> Great, I'll take care of that PR

Nice

File: site/docs/guide/files.md [hunk 7]
```diff
@@ -181,12 +181,20 @@ new InputFile("/path/to/file");
...
+new InputFile(await Bun.file("/path/to/file").bytes());
+```
+
 :::
 
 #### Uploading Raw Binary Data
 
 You can also send a `Buffer` object, or an iterator that yields `Buffer` objects.
-On Deno, you can send `Blob` objects, too.
+On Deno and Bun you can send `Blob` objects, too.
```
Comment (KnorpelSenf) [inline site/docs/guide/files.md:197]:
```suggestion
On Deno and Bun, you can send `Blob` objects, too.
```

File: site/docs/guide/files.md [hunk 8]
```diff
@@ -181,12 +181,20 @@ new InputFile("/path/to/file");
 new InputFile(await Deno.open("/path/to/file"));
 ```
 
+```ts [Bun]
+// Send a local file.
+new InputFile("/path/to/file");
+
+// Send a `BunFile` instance.
+new InputFile(await Bun.file("/path/to/file").bytes());
```
Comment (KnorpelSenf) [inline site/docs/guide/files.md:189]: Why would
```suggestion
new InputFile(await Bun.file("/path/to/file"));
```
not work?

Comment (MasedMSD) [reply site/docs/guide/files.md:189]: Unfortunately no, we will have to fix this at the level of the grammY itself and add support for BunFile.
image

Comment (KnorpelSenf) [reply site/docs/guide/files.md:189]: Then their docs are wrong: https://bun.sh/docs/api/file-io#reading-files-bun-file

They say that they conform to `Blob` so that means that `Bun.file("/tmp/file") instanceof Blob` should be `true`.

Comment (KnorpelSenf) [reply site/docs/guide/files.md:189]: Their reference is also wrong because their example states that `.bytes()` exists but their reference does not list it. Or is that implicit via `Blob` which they don't seem to inherit? Also, why did they list `.stream()` then?

Comment (KnorpelSenf) [reply site/docs/guide/files.md:189]: But well, I guess that you are right that we cannot perform this change

**Comment by KnorpelSenf** [comment]
Closing due to inactivity. I will add bun-specific docs myself for the 2.0 rewrite of the docs.
